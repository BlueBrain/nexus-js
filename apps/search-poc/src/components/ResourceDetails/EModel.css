
.small-table td {
  padding: 4px 16px !important;
}

