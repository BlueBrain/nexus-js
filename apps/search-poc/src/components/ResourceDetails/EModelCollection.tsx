import React from 'react';
import { Table } from 'antd';

export type Emodel = {
  id: string;
  self: string;
  name: string;
  description: string;
};

const emodelTableColumns = [
  {
    title: 'Name',
    render: (_: any, emodel: Emodel) => (
      <a key={emodel.id} href={`/resources?self=${emodel.self}`}>
        {emodel.name}
      </a>
    ),
    key: 'name',
  },
  {
    title: 'Description',
    dataIndex: 'description',
    key: 'description',
  },
];

const EModelCollectionDetails: React.FunctionComponent<{
  resource: {
    emodels: Emodel[];
  };
}> = props => {
  return (
    <div className="emodel-collection-details">
      <Table
        rowKey="id"
        className="small-table"
        columns={emodelTableColumns}
        dataSource={props.resource.emodels}
        title={() => 'EModels'}
        size="middle"
        bordered
        pagination={false}
      />
    </div>
  );
};

export default EModelCollectionDetails;
