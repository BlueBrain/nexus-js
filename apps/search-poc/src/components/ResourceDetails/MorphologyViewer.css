
.morphology {
  height: 400px;
}
