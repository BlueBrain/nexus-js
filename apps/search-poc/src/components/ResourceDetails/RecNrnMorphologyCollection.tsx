import React from 'react';
import { Table } from 'antd';

export type Morph = {
  id: string;
  self: string;
  name: string;
  description: string;
};

const morphTableColumns = [
  {
    title: 'Name',
    render: (_: any, morph: Morph) => (
      <a key={morph.id} href={`/resources?self=${morph.self}`}>
        {morph.name}
      </a>
    ),
    key: 'name',
  },
  {
    title: 'Description',
    dataIndex: 'description',
    key: 'description',
  },
];

const RecNrnMorphologyCollection: React.FunctionComponent<{
  resource: {
    reconstructedcells: Morph[];
  };
}> = props => {
  return (
    <div className="emodel-collection-details">
      <Table
        rowKey="id"
        title={() => 'Reconstructed neuron morphologies'}
        className="small-table"
        columns={morphTableColumns}
        dataSource={props.resource.reconstructedcells}
        size="middle"
        bordered
      />
    </div>
  );
};

export default RecNrnMorphologyCollection;
