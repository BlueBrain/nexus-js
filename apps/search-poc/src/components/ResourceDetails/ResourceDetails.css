.resource-details {
  width: 980px;
  margin: 0 auto;
  display: flex;
  justify-content: space-between;
}

.metadata-container {
}

.metadata-container > ul {
  padding: 0;
  margin: 0;
}
.metadata-container > .name {
  font-size: 3em;
  font-weight: lighter;
}

.layers-container {
  margin-bottom: 1em;
}
