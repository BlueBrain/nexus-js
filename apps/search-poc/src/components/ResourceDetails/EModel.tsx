
import React, { ReactNode } from 'react';
import sortBy from 'lodash/sortBy';
import { Table, Collapse } from 'antd';

import { Resource } from '@bbp/nexus-sdk';
import { MINDSResource, EModelResource } from '../../containers/ResourceDetails/types';


const { Panel } = Collapse;
const protocolR = /^(?:_?\.)(.*)\.(\w+)\.(\w+)\.(\w+)$/;

const paramColumns = [{
  title: 'Parameter',
  dataIndex: 'param',
  key: 'param',
}, {
  title: 'Value',
  dataIndex: 'val',
  key: 'val',
}];


/**
 * Parses fitness raw parameter name into:
 * * protocol (formatted by replacing underscores with spaces)
 * * recTarget
 * * recType
 * * param
 *
 * @param rawParamName [string] Fitness raw parameter name
 *
 * @example
 * parseFitnessRawParamName('_.APWaveform_300.soma.v.AHP_depth_abs');
 * {
 *   protocol: 'APWaveform_300',
 *   recTarget: 'soma',
 *   recType: 'v',
 *   param: 'AHP_depth_abs'
 * }
 */
export function parseFitnessRawParamName(rawParamName: string): {
  protocol: string;
  recTarget: string;
  recType: string;
  param: string;
} {
  const [, rawProtocolName, recTarget, recType, param] = rawParamName
    .match(protocolR) as string[];

  const protocol = rawProtocolName.replace(/_/g, ' ');

  return { protocol, recTarget, recType, param };
}

function createUniqColumnRenderer(
  dataSource: object[],
  dataKey: string,
  colKey: string
): (value: any, row: any, index: number) => ReactNode | undefined {
  return (value, row, index) => {
    const obj = { children: value[dataKey], props: { rowSpan: 1 } };
    const isRendered = dataSource
      .slice(0, index)
      .some(entry => entry[colKey] === value[colKey]);

    obj.props.rowSpan = isRendered
      ? 0
      : dataSource.filter(entry => entry[colKey] === value[colKey]).length;

    return obj;
  };
};

function createFitnessDataSourceEntry([paramRaw, val]) {
  const {
    protocol,
    recTarget,
    recType,
    param
  } = parseFitnessRawParamName(paramRaw);

  const key = paramRaw;

  return { key, protocol, recTarget, recType, param, val };
};

const EModelDetails: React.FunctionComponent<{
  resource: Resource & MINDSResource & EModelResource;
}> = props => {
  const { score, seed, fitness, params } = props.resource;

  const fitnessDSUnsorted = Object.entries(fitness)
    .map(createFitnessDataSourceEntry);

  const fitnessDataSource = sortBy(fitnessDSUnsorted, ['parameter', 'recTarget', 'protocol']);

  const fitnessColumns = [{
    title: 'Protocol',
    render: createUniqColumnRenderer(fitnessDataSource, 'protocol', 'protocol'),
    key: 'protocol',
  }, {
    title: 'Target',
    render: createUniqColumnRenderer(fitnessDataSource, 'recTarget', 'protocol'),
    key: 'recTarget',
  }, {
    title: 'Type',
    render: createUniqColumnRenderer(fitnessDataSource, 'recType', 'protocol'),
    key: 'recType'
  }, {
    title: 'Parameter',
    dataIndex: 'param',
    key: 'param',
  }, {
    title: 'Value',
    dataIndex: 'val',
    key: 'val',
  }];

  const paramDataSource = Object.entries(params)
    .map(([key, val]) => ({ key, param: key, val }));

  return (
    <div className="emodel-details">
      <p>Seed: {seed}</p>
      <p>Score: {score}</p>

      <Collapse>
        <Panel header="Fitness" key="fitness">
          <Table
            className="small-table"
            dataSource={fitnessDataSource}
            columns={fitnessColumns}
            pagination={false}
            size="middle"
            bordered
          />
        </Panel>
        <Panel header="Optimized parameters" key="params">
          <Table
            className="small-table"
            dataSource={paramDataSource}
            columns={paramColumns}
            pagination={false}
            size="middle"
            bordered
          />
        </Panel>
      </Collapse>
    </div>
  );
};

export default EModelDetails;
