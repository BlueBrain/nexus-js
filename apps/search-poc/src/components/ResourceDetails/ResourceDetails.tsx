import * as React from 'react';
import { Tag, Icon } from 'antd';

import './ResourceDetails.css';
import { camelCaseToLabelString } from '../../utils';

const ResourceDetails: React.FunctionComponent<{
  id: string;
  types: string[];
  name: string;
  description: string;
  brainRegion: {
    id: string;
    label: string;
  };
  layers: {
    id: string;
    label: string;
  }[];
  species: {
    id: string;
    label: string;
  };
  strain: {
    id: string;
    label: string;
  };
  uploadedAt: string;
  uploadedBy: string;
  className: string;
}> = props => {
  const NotFound = ({ label }: { label: string }) => {
    return (
      <p>
        <Icon type="warning" />
        This dataset does not have <b>{label}</b> information
      </p>
    );
  };
  return (
    <div className={`resource-details ${props.className}`}>
      <div className="metadata-container">
        <h1 className="name">{props.name}</h1>

        <h2>
          {props.types.map(type => (
            <em>{camelCaseToLabelString(type)}</em>
          ))}
        </h2>

        <p>
          uploaded {props.uploadedAt} by <b>{props.uploadedBy}</b>
        </p>

        <div className="brain-region">
          {props.brainRegion.label ? (
            <a
              target="_blank"
              rel="noopener noreferrer"
              href={props.brainRegion.id}
            >
              {props.brainRegion.label}
            </a>
          ) : (
            <NotFound label="Brain Region" />
          )}
        </div>

        {props.layers.length ? (
          <div className="layers-container">
            {props.layers.map(layer => (
              <Tag key={layer.id} color="blue">
                <a target="_blank" rel="noopener noreferrer" href={layer.id}>
                  {layer.label}
                </a>
              </Tag>
            ))}
          </div>
        ) : null}

        <p>
          {props.species.label ? (
            <a
              target="_blank"
              rel="noopener noreferrer"
              href={props.species.id}
            >
              {props.species.label}
            </a>
          ) : (
            <NotFound label="Species" />
          )}
        </p>

        <p>
          {props.strain.label ? (
            <a target="_blank" rel="noopener noreferrer" href={props.strain.id}>
              {props.strain.label}
            </a>
          ) : (
            <NotFound label="Strain" />
          )}
        </p>

        <p className="descripion">
          {props.description ? (
            props.description
          ) : (
            <NotFound label="description" />
          )}
        </p>
      </div>
      <div className="module-container">{props.children}</div>
    </div>
  );
};

export default ResourceDetails;
