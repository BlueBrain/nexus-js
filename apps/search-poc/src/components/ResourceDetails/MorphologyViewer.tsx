import * as React from 'react';
import { Card, Spin, Empty } from 'antd';
import './MorphologyViewer.css';

const MorphologyViewer: React.FunctionComponent<{
  loading: boolean;
  error: Error | null;
  stage?: string;
  morphology: any;
  ref: React.RefObject<HTMLDivElement>;
}> = ({ loading, error, ref, stage }) => {
  return (
    <Spin spinning={loading} tip={stage}>
      <Card hoverable className="morphology-viewer">
        <div className={`viewer ${!loading && !error && '-loaded'}`} ref={ref}>
          {error && <Empty>{error.message}</Empty>}
        </div>
      </Card>
    </Spin>
  );
};

export default MorphologyViewer;
