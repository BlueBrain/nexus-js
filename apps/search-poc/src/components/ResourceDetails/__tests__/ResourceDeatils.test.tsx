import ResourceDetails  from '../ResourceDetails';
import * as renderer from "react-test-renderer";
import React from "react";
import { shallow } from "enzyme";



it('ResourceDetails renders correctly', () => {
    const props = {
        strain: {
            "@id": "http://purl.obolibrary.org/obo/RS_0001833",
            "label": "Wistar Han"
          }, 
        species: {
            "@id": "http://purl.obolibrary.org/obo/NCBITaxon_10116",
            "label": "Rattus norvegicus"
          }, 
        types: ["EModel", "Entity"], 
        layers: [
            {
              "@id": "http://purl.obolibrary.org/obo/UBERON_0005390",
              "label": "cortical layer I"
            },
            {
              "@id": "http://purl.obolibrary.org/obo/UBERON_0005392",
              "label": "cortical layer III"
            },
            {
              "@id": "http://purl.obolibrary.org/obo/UBERON_0005394",
              "label": "cortical layer V"
            },
            {
              "@id": "http://purl.obolibrary.org/obo/UBERON_0005395",
              "label": "cortical layer VI"
            },
            {
              "@id": "http://purl.obolibrary.org/obo/UBERON_0005391",
              "label": "cortical layer II"
            },
            {
              "@id": "http://purl.obolibrary.org/obo/UBERON_0005393",
              "label": "cortical layer IV"
            }
          ], 
        brainRegion: {
            "@id": "http://purl.obolibrary.org/obo/UBERON_0008933",
            "label": "primary somatosensory cortex"
        }, 
        resource: { 
            score: 212.6819067869392, 
            seed: "1", 
            fitness:  {
                "_.APWaveform_300.soma.v.AHP_depth_abs": 1.4844549844325083,
                "_.APWaveform_300.soma.v.AP1_amp": 1.2938048460056084,
                "_.APWaveform_300.soma.v.AP_amplitude": 1.5379618971924824,
                "_.APWaveform_300.soma.v.AP_duration_half_width": 1.0126582278483314
            }, 
            params: {
                "decay_CaDynamics_DC0.axonal": 321.1455936884192,
                "decay_CaDynamics_DC0.somatic": 359.9059029669804,
                "e_pas.all": -70.84079169935472,
                "gCa_HVAbar_Ca_HVA2.axonal": 0.00018988503473868083
            }
        }
    };

    const result = shallow(<ResourceDetails {...props} />);
    expect(result).toMatchSnapshot();
});