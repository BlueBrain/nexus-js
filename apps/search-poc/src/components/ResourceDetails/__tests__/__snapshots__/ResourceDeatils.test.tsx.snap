// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResourceDetails renders correctly 1`] = `
<div
  className="ResourceDetails"
>
  <div
    className="metadata-container"
  >
    <h1
      className="name"
    />
    <h3>
      Type:  
      <Tag
        closable={false}
        key="EModel"
      >
        EModel
      </Tag>
      <Tag
        closable={false}
        key="Entity"
      >
        Entity
      </Tag>
    </h3>
    <p>
      Brain region:  
      <a
        rel="noopener noreferrer"
        target="_blank"
      >
        primary somatosensory cortex
      </a>
    </p>
    <p>
      Species:  
      <a
        rel="noopener noreferrer"
        target="_blank"
      >
        Rattus norvegicus
      </a>
    </p>
    <p>
      Strain:  
      <a
        rel="noopener noreferrer"
        target="_blank"
      >
        Wistar Han
      </a>
    </p>
    <div
      className="layers-container"
    >
      Layers:  
      <Tag
        closable={false}
        color="blue"
      >
        <a
          rel="noopener noreferrer"
          target="_blank"
        >
          cortical layer I
        </a>
      </Tag>
      <Tag
        closable={false}
        color="blue"
      >
        <a
          rel="noopener noreferrer"
          target="_blank"
        >
          cortical layer III
        </a>
      </Tag>
      <Tag
        closable={false}
        color="blue"
      >
        <a
          rel="noopener noreferrer"
          target="_blank"
        >
          cortical layer V
        </a>
      </Tag>
      <Tag
        closable={false}
        color="blue"
      >
        <a
          rel="noopener noreferrer"
          target="_blank"
        >
          cortical layer VI
        </a>
      </Tag>
      <Tag
        closable={false}
        color="blue"
      >
        <a
          rel="noopener noreferrer"
          target="_blank"
        >
          cortical layer II
        </a>
      </Tag>
      <Tag
        closable={false}
        color="blue"
      >
        <a
          rel="noopener noreferrer"
          target="_blank"
        >
          cortical layer IV
        </a>
      </Tag>
    </div>
    <p
      className="descripion"
    >
      Description: 
    </p>
  </div>
  <div
    className="module-container"
  />
</div>
`;
